proposals/0000-arraybuffer-support-in-turbomodules.md

```diff
@@ -0,0 +1,381 @@
+---
+title: ArrayBuffer support in TurboModules
+author:
+  - Kamil Paradowski <kamil.paradowski@callstack.com>
+date: 10.10.2025
+---
+
+# RFC0000: ArrayBuffer support in TurboModules
+
+## Summary
+
+This RFC outlines adding first-class `ArrayBuffer` support to TurboModules to enable zero-copy binary data exchange between JavaScript and native modules.
+
+Codegen and TurboModules currently lack efficient binary data handling, forcing developers to rely on workarounds, such as using the `CodegenTypes.UnsafeObject` type or inefficient conversion to `base64`, when working with binary-heavy use cases. This feature addresses performance bottlenecks in data-intensive applications. By providing first-class `ArrayBuffer` support, developers will be able to seamlessly pass large binary payloads between JavaScript and native code without the overhead of data serialization and copying. This enhancement will unlock new possibilities for creating responsive, high-performance React Native applications across various domains including media processing, real-time communication, and AI/ML applications that require efficient handling of tensors and model data.
+
+## Basic example
+
+Developers should be able to define a TypeScript spec as follows:
+
+```ts
+export interface Spec extends TurboModule {
+  getBuffer(): ArrayBuffer;
+  processBuffer(buffer: ArrayBuffer): void;
+}
+```
+
+## Motivation
+
+TurboModules currently lack a first-class way to represent `ArrayBuffer` end-to-end in Codegen, which forces developers to rely on copies, ad-hoc platform bridges, global helpers, or external libraries. This hurts performance for binary-heavy use cases such as media data or ML tensors, and it increases implementation complexity. The expected outcome is a cross-platform contract that lets JS and native pass binary data with minimal copying. Codegen should be able to generate working code for the `ArrayBuffer` type on every platform.
+
+For example, several important use cases are currently difficult to implement efficiently while working with TurboModules:
+
+- **Real-time media streaming**: A native video decoder could stream frames directly to a JavaScript-based player component. Without zero-copy `ArrayBuffer`s, each frame would need to be copied, leading to significant performance overhead and potential frame drops.
```

Comment:
I think great example here would be to use Blob Manager as an example (and likely first candidate to migrate over, once this lands). 

Broadly speaking working with binary data in general.